From ed31d6a309d9900271c05157062a72cb02149c16 Mon Sep 17 00:00:00 2001
From: Andy Grover <agrover@redhat.com>
Date: Wed, 26 Nov 2014 19:53:17 -0500
Subject: [ib_isert] Add max_send_sge=2 minimum for control PDU responses

Message-id: <1417031598-5408-2-git-send-email-agrover@redhat.com>
Patchwork-id: 100976
O-Subject: [RHEL7.1 PATCH e-stor 1/2] ib_isert: Add max_send_sge=2 minimum for control PDU responses
Bugzilla: 1166314
RH-Acked-by: Chris Leech <cleech@redhat.com>
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>

From: Or Gerlitz <ogerlitz@mellanox.com>

This patch adds a max_send_sge=2 minimum in isert_conn_setup_qp()
to ensure outgoing control PDU responses with tx_desc->num_sge=2
are able to function correctly.

This addresses a bug with RDMA hardware using dev_attr.max_sge=3,
that in the original code with the ConnectX-2 work-around would
result in isert_conn->max_sge=1 being negotiated.

Originally reported by Chris with ocrdma driver.

Reported-by: Chris Moore <Chris.Moore@emulex.com>
Cc: <stable@vger.kernel.org> # 3.10+
Signed-off-by: Or Gerlitz <ogerlitz@mellanox.com>
Signed-off-by: Nicholas Bellinger <nab@linux-iscsi.org>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/infiniband/ulp/isert/ib_isert.c b/drivers/infiniband/ulp/isert/ib_isert.c
index 0aa33cd..70f56ab 100644
--- a/drivers/infiniband/ulp/isert/ib_isert.c
+++ b/drivers/infiniband/ulp/isert/ib_isert.c
@@ -107,9 +107,12 @@ isert_conn_setup_qp(struct isert_conn *isert_conn, struct rdma_cm_id *cma_id)
  attr.cap.max_recv_wr = ISERT_QP_MAX_RECV_DTOS;
  /*
   * FIXME: Use devattr.max_sge - 2 for max_send_sge as
-  * work-around for RDMA_READ..
+  * work-around for RDMA_READs with ConnectX-2.
+  *
+  * Also, still make sure to have at least two SGEs for
+  * outgoing control PDU responses.
   */
- attr.cap.max_send_sge = devattr.max_sge - 2;
+ attr.cap.max_send_sge = max(2, devattr.max_sge - 2);
  isert_conn->max_sge = attr.cap.max_send_sge;
 
  attr.cap.max_recv_sge = 1;
-- 
1.7.1